File: Clone Graph.java
/**
 * Definition for undirected graph.
 * class UndirectedGraphNode {
 *     int label;
 *     List<UndirectedGraphNode> neighbors;
 *     UndirectedGraphNode(int x) { label = x; neighbors = new ArrayList<UndirectedGraphNode>(); }
 * };
 */
public class Solution {
    public UndirectedGraphNode clone(UndirectedGraphNode node, Map<Integer, UndirectedGraphNode> cache) {
        if (node == null)
            return null;
        if (!cache.containsKey(node.label)) {
            UndirectedGraphNode cloned = new UndirectedGraphNode(node.label);
            cache.put(node.label, cloned);
            for (UndirectedGraphNode neighbor : node.neighbors) {
                cloned.neighbors.add(clone(neighbor, cache));
            }
        }
        return cache.get(node.label);
    }
    public UndirectedGraphNode cloneGraph(UndirectedGraphNode node) {
        Map<Integer, UndirectedGraphNode> cache = new HashMap<Integer, UndirectedGraphNode>();
        return clone(node, cache); 
    }
}

File: Clone Graph.cpp
/**
 * Definition for undirected graph.
 * struct UndirectedGraphNode {
 *     int label;
 *     vector<UndirectedGraphNode *> neighbors;
 *     UndirectedGraphNode(int x) : label(x) {};
 * };
 */
class Solution {
public:
    UndirectedGraphNode* clone(UndirectedGraphNode *node, map<int, UndirectedGraphNode*>& cache) {
        if (!node)
            return 0;
        if (cache.find(node->label) == cache.end()) {
            UndirectedGraphNode* cloned = new UndirectedGraphNode(node->label);
            cache[node->label] = cloned;
            for (int i = 0; i < node->neighbors.size(); i++) {
                cloned->neighbors.push_back( clone(node->neighbors[i], cache) );
            }
        }
        return cache[node->label];
    }
    UndirectedGraphNode *cloneGraph(UndirectedGraphNode *node) {
        map<int, UndirectedGraphNode*> cache;
        return clone(node, cache);
    }
};